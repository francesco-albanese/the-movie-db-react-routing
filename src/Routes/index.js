import { isEmpty } from 'lodash-es'
import React from 'react'
import { Switch } from 'react-router-dom'

import { renderRoutes } from '#utils'

export default class Routes extends React.Component {

  state = {
    routes: [],
    activeLocale: {}
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      activeLocale,
      pages,
      routesConfig
    } = nextProps

    if (activeLocale.code !== prevState.activeLocale.code) {
      return {
        activeLocale,
        routes: renderRoutes({
          activeLocale,
          pages,
          routesConfig
        })
      }
    }

    return null
  }

  render() {
    const { routes } = this.state

    return !isEmpty(routes)
      ? <Switch>{ routes }</Switch>
      : null
  }
}