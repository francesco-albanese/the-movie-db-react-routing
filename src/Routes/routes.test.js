import Routes from '.'
import { renderRoutes } from '#utils'
import { 
  activeLocale, 
  pages, 
  routesConfig 
} from './__fixtures__/routes.fixtures'

describe('Routes', () => {

  it('renders', () => {
    expect(shallow(
      <Routes 
        activeLocale={ activeLocale }
        pages={ pages }
        routesConfig={ routesConfig } />)
    ).toMatchSnapshot()
  })

  describe.only('renderRoutes', () => {
    it('returns empty array if missing configs', () => {
      const result = renderRoutes()
      expect(result).toEqual([])
    })

    it('returns only 404 if page reference does not match with route.reference', () => {
      const newRoutesConfig = routesConfig
        .filter(({ path }) => path === null)
        .concat({ 
          reference: 'fake-reference',
          exact: true,
          component: (props) => props.children
        })

      const result = renderRoutes({ activeLocale, pages, routesConfig: newRoutesConfig })
      expect(result[ 0 ].key).toEqual('four-oh-four')
    })

    it('returns list of routes', () => {
      const result = renderRoutes({ activeLocale, pages, routesConfig })
      expect(result).toHaveLength(4)
      expect(result).toMatchSnapshot()
    })
  })
})