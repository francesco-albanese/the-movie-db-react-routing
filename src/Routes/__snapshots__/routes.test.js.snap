// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes renderRoutes returns list of routes 1`] = `
Array [
  <Route
    component={[Function]}
    exact={true}
    path="/en"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/en/favourites"
  />,
  <Route
    component={[Function]}
    exact={true}
    path="/en/movie/:movieid"
  />,
  <Route
    component={[Function]}
    exact={true}
    path={null}
  />,
]
`;

exports[`Routes renders 1`] = `
<Switch>
  <Route
    component={[Function]}
    exact={true}
    key="home-page"
    path="/en"
  />
  <Route
    component={[Function]}
    exact={true}
    key="favourites-page"
    path="/en/favourites"
  />
  <Route
    component={[Function]}
    exact={true}
    key="movie-details-page"
    path="/en/movie/:movieid"
  />
  <Route
    component={[Function]}
    exact={true}
    key="four-oh-four"
    path={null}
  />
</Switch>
`;
